streams/src/commonMain/kotlin/com/mirego/trikot/streams/reactive/processors/TakeUntilProcessor.kt

```diff
@@ -0,0 +1,49 @@
+package com.mirego.trikot.streams.reactive.processors
+
+import com.mirego.trikot.foundation.concurrent.atomic
+import com.mirego.trikot.streams.reactive.StreamsProcessorException
+import org.reactivestreams.Publisher
+import org.reactivestreams.Subscriber
+
+typealias TakeUntilProcessorPredicate<T> = (T) -> Boolean
+
+class TakeUntilProcessor<T>(
+    parentPublisher: Publisher<T>,
+    private val predicate: TakeUntilProcessorPredicate<T>
+) : AbstractProcessor<T, T>(parentPublisher) {
+
+    override fun createSubscription(subscriber: Subscriber<in T>): ProcessorSubscription<T, T> =
+        TakeUntilProcessorSubscription(subscriber, predicate)
+
+    class TakeUntilProcessorSubscription<T>(
+        subscriber: Subscriber<in T>,
+        private val predicate: TakeUntilProcessorPredicate<T>
+    ) : ProcessorSubscription<T, T>(subscriber) {
+
+        private var hasCompleted: Boolean by atomic(false)
+
+        override fun onNext(t: T, subscriber: Subscriber<in T>) {
+            if (hasCompleted) return
+            subscriber.onNext(t)
```

Comment:
I feel like this should be run at line 35 (after evaluating the predicate) in order to catch possible errors before actually propagating the current value. Would that make more sense? I'll check the Rx spec to see if this is specified.

Reply:
According to the [java implementation](https://github.com/ReactiveX/RxJava/blob/4749ab3879dcfb4afcf84edcade6141a7221c194/src/main/java/io/reactivex/rxjava3/internal/operators/observable/ObservableTakeUntilPredicate.java#L64) the object is passed downstream before testing the predicate. We probably want to emit the object that caused an error before becoming in an error state.

Reply:
@CharlesMcD Nice 👍🏻 Sounds good then!